Userland/Libraries/LibWeb/Fetch/FetchMethod.cpp

```diff
@@ -0,0 +1,179 @@
+    (void)global_object;
+
+    // 7. Let responseObject be null.
+    JS::Handle<Response> response_object_handle;
+
+    // 8. Let relevantRealm be this’s relevant Realm.
+    // NOTE: This assumes that the running execution context is for the fetch() function call.
+    auto& relevant_realm = HTML::relevant_realm(*vm.running_execution_context().function);
+
+    // 9. Let locallyAborted be false.
+    // NOTE: This lets us reject promises with predictable timing, when the request to abort comes from the same thread
+    //       as the call to fetch.
+    auto locally_aborted = Fetching::RefCountedFlag::create(false);
+
+    // 10. Let controller be null.
+    RefPtr<Infrastructure::FetchController> controller;
+
+    // 11. Add the following abort steps to requestObject’s signal:
+    request_object->signal()->add_abort_algorithm([&vm, locally_aborted, request, controller, promise_capability_handle = JS::make_handle(*promise_capability), request_object_handle = JS::make_handle(*request_object), response_object_handle]() mutable {
+        dbgln_if(WEB_FETCH_DEBUG, "Fetch: Request object signal's abort algorithm called");
+
+        auto& promise_capability = *promise_capability_handle;
+        auto& request_object = *request_object_handle;
+        auto& response_object = *response_object_handle;
+
+        // 1. Set locallyAborted to true.
+        locally_aborted->set_value(true);
+
+        // 2. Abort the fetch() call with p, request, responseObject, and requestObject’s signal’s abort reason.
+        abort_fetch(vm, promise_capability, request, response_object, request_object.signal()->reason());
+
+        // 3. If controller is not null, then abort controller with requestObject’s signal’s abort reason.
+        if (!controller.is_null())
+            controller->abort(vm, request_object.signal()->reason());
+    });
+
+    // 12. Set controller to the result of calling fetch given request and processResponse given response being these
+    //     substeps:
+    auto process_response = [&vm, locally_aborted, promise_capability, request, response_object_handle, &relevant_realm](NonnullRefPtr<Infrastructure::Response> response) mutable {
+        // 1. If locallyAborted is true, terminate these substeps.
+        if (locally_aborted->value())
+            return;
+
+        // 2. If response’s aborted flag is set, then:
+        if (response->aborted()) {
+            // FIXME: 1. Let deserializedError be the result of deserialize a serialized abort reason given controller’s
+            //           serialized abort reason and relevantRealm.
+            auto deserialized_error = JS::js_undefined();
+
+            // 2. Abort the fetch() call with p, request, responseObject, and deserializedError.
+            abort_fetch(vm, promise_capability, request, response_object_handle.cell(), deserialized_error);
+            return;
```

Comment:
This doesn't say `terminate these substeps` but that could very well be a spec issue

Reply:
oh fun fact, it used to say that until very recently - I actually started implementing this before they merged https://github.com/whatwg/fetch/commit/c9039b97a502b1d5fea46b55b9166762c3479837#diff-b4827ee42aceaf9dc3e7216f72262d81c32217e9996fb70a657f259e3ac92604L7818-R7872

I believe it's harmless to omit from a spec perspective as aborting will reject the promise, and then either rejecting or resolving the already fulfilled promise in the following steps is a no-op.